### DISCUSSION TITLE
DiceMetric gives weird results

### DISCUSSION THREAD

**Discussion by franciskasara** [Project-MONAI/MONAI#7427] [Q&A] [answered]
Hi! 

I am training a MONAI `model =SegResNet( out_channels=2)` to a segmentation task. I have 2 classes, foreground and background. The groundtruth segmentations/labels are 1 channel images. 
I use the MONAI `DiceLoss(softmax=True, include_background=False, to_onehot_y=True)` which seems to work, it decreases, and the predictions look decent. (`include_background`  is False, because the background is much bigger as the foreground.) However I just can't seem to figure out how to use `DiceMetric(include_background=False, reduction="mean", get_not_nans=False)`.

It either gives an error, or 1 or 0 or numbers above 1. 
I read all the tutorials I could find, and try to copy those with no success.... 

I use the loss like this:
` vloss = loss_fn(voutputs, vlabels)`

In the tutorials "AsDiscreted" was a common step. I do not want to use MONAI transforms, so this is what I tried:
```
voutputs_bin=voutputs
voutputs_bin=torch.nn.functional.softmax(voutputs_bin,dim=1)
#voutputs_bin=torch.argmax(voutputs_bin,dim=1)
voutputs_bin = torch.nn.functional.one_hot(voutputs_bin.to(torch.int64), num_classes=-1)
#voutputs_bin=voutputs_bin[:,1:,:,:]
#voutputs_bin=torch.nn.functional.threshold(voutputs_bin,0.5,1)
vlabels_bin=torch.nn.functional.one_hot(vlabels.to(torch.int64), num_classes=-1)

metric(y_pred=voutputs_bin, y=vlabels_bin) 
```

I included the commented lines, because this, and different combinations of this is what I tried. 
After the validation loop I do:
```
vmetric=metric.aggregate().item()
print(f' dice metric: {vmetric}')
metric.reset()
```

Could you please tell and explain me, how should I properly use the DiceMetric?

**Comment by KumoLiu** [comment]
Hi @franciskasara, could you please paste the error message?
And you can refer to the usage here.
https://github.com/Project-MONAI/tutorials/blob/3093dec599f04657a7ae6c7a9c97eba5c910259a/modules/compute_metric.py#L110-L114
Hope it helps, thanks.

**Reply by franciskasara** [reply]
I have already seen this, and no, it did not help at all. I have tried to use thresholding, like they do in your linked example.

My main concern is NOT the error, it is the metric not working properly. And I'd really like to focus this discussion on how to use it in this case, and the explanation why.
 
(But when I got the error it was on the aggregating part:  RuntimeError: stack expects each tensor to be equal size, but got [1] at entry 0 and [255] at entry 384)

**Reply by KumoLiu** [reply]
Perhaps posting a small snippet of code that reproduces the issue will make it easier for me to spot the problem. Thanks.

**Comment by franciskasara** [comment, marked as answer]
Okay, I found the problem. 

- In the threshold version the problem was that 
`voutputs_bin=torch.nn.functional.threshold(voutputs_bin,threshold=0.5,value=1)`
does not binarize the tensor. It sets everything under 0.5 to 1, and leaves the rest of it untouched. 


- In the argmax version the problem was that after
` voutputs_bin=torch.argmax(voutputs_bin,dim=1)`  
I did not set the `num_classes` parameter in `DiceMetric` to 2, so it messed it up:
"_num_classes – number of input channels (always including the background). **When this is None, y_pred.shape[1]**_"


This is how it works properly: (although it is not going to be 1-DiceLoss.)
```
voutputs_bin=torch.nn.functional.softmax(voutputs,dim=1)
voutputs_bin=voutputs_bin>=0.5

metric(y_pred=voutputs_bin, y=vlabels.to(torch.int64))

```

**Reply by KumoLiu** [reply]
BTW, you can use `AsDiscrete` in MONAI to convert data into discrete values.
https://github.com/Project-MONAI/MONAI/blob/df4cf922e048aeba2207641654e2b61995e1c055/monai/transforms/post/array.py#L132
